src/stream/src/executor/stream_reader.rs

```diff
@@ -64,38 +67,50 @@ impl SourceReaderStream {
         }
     }
 
-    /// Convert this reader to a stream.
-    pub fn new(
+    /// Construct a `ReaderStreamWithPause` with one stream receiving streaming messages and the
+    /// other receiving data only.
+    pub fn new_with_message_stream(
+        message_stream: ExecutorMessageStream,
+        data_stream: BoxSourceWithStateStream,
+    ) -> Self {
+        let message_stream_arm = message_stream.map_ok(Either::Left).boxed();
+        let data_stream_arm = Self::data_stream(data_stream).map_ok(Either::Right).boxed();
+        let inner = Self::new_inner(message_stream_arm, data_stream_arm);
+        Self {
+            inner,
+            paused: false,
+        }
+    }
+
+    /// Construct a `ReaderStreamWithPause` with one stream receiving barriers only and the other
+    /// receiving data only.
+    pub fn new_with_barrier_receiver(
         barrier_receiver: UnboundedReceiver<Barrier>,
-        source_stream: BoxSourceWithStateStream,
+        data_stream: BoxSourceWithStateStream,
     ) -> Self {
+        let barrier_receiver_arm = Self::barrier_receiver(barrier_receiver)
```

Comment:
Call we directly call `new_with_message_stream` with `Self::barrier_receiver(barrier_receiver)` here?